src/langchain_google_firestore/document_loader.py

```diff
@@ -0,0 +1,217 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import itertools
+from typing import TYPE_CHECKING, Any, Iterator, List, Optional
+
+import more_itertools
+from google.cloud import firestore  # type: ignore
+from google.cloud.firestore import DocumentReference  # type: ignore
+from langchain_community.document_loaders.base import BaseLoader
+from langchain_core.documents import Document
+
+from .utility.document_converter import DocumentConverter
+
+USER_AGENT_LOADER = "langchain-google-firestore-python:document_loader"
+USER_AGENT_SAVER = "langchain-google-firestore-python:document_saver"
+WRITE_BATCH_SIZE = 500
+TYPE = "firestore_type"
+
+if TYPE_CHECKING:
+    from google.cloud.firestore import Client, CollectionGroup, DocumentReference, Query
+
+
+class FirestoreLoader(BaseLoader):
+    def __init__(
+        self,
+        source: Query | CollectionGroup | DocumentReference | str,
+        page_content_fields: List[str] = [],
+        metadata_fields: List[str] = [],
+        client: Client = None,
+    ) -> None:
+        """Document Loader for Google Cloud Firestore.
+        Args:
+            source: The source to load the documents. It can be an instance of Query,
+                CollectionGroup, DocumentReference or the single `/`-delimited path to
+                a Firestore collection.
+            page_content_fields: The document field names to write into the `page_content`.
+                If an empty or None list is provided all fields will be written into
+                `page_content`. When only one field is provided only the value is written.
+            metadata_fields: The document field names to write into the `metadata`.
+                By default it will write all fields that are not in `page_content` into `metadata`.
+            client: Client for interacting with the Google Cloud Firestore API.
+        """
+        self.client = client or firestore.Client()
+        client_agent = self.client._client_info.user_agent
+        if not client_agent:
+            self.client._client_info.user_agent = USER_AGENT_LOADER
+        elif USER_AGENT_LOADER not in client_agent:
+            self.client._client_info.user_agent = " ".join(
+                [client_agent, USER_AGENT_LOADER]
+            )
```

Comment:
nit: move this into a function -- something like "client_w_user_agent", when we can use it like:

client = client_w_user_agent(client)